### ISSUE TITLE
Need a PluginErrorHandler interface or something that telegraf can callback when it fails to poll for metrics.

### ISSUE THREAD

**Issue by batistado** [influxdata/telegraf#10245]
## Feature Request

Need a PluginErrorHandler interface or something that telegraf can callback when it fails to poll for metrics. 
For more context please check out this thread: https://influxcommunity.slack.com/archives/CH7C56X4Z/p1638920766176700

### Proposal:

A new interface something like this:
So an interface something like this:
```
type PluginErrorHandler interface {
    OnError(err error)
}
```
would be really useful. 
https://influxcommunity.slack.com/archives/CH7C56X4Z/p1638984531179900?thread_ts=1638920766.176700&cid=CH7C56X4Z

### Current behavior:
On this line: https://github.com/influxdata/telegraf/blob/34ad5aa137f0d845d79826839379cb543c3879c5/agent/agent.go#L451 I see the error being added to the accumulator and the accumulator just logs the error.

### Desired behavior:
The accumulator should also call the OnError method of the PluginErrorHandler interface in addition to logging the error.

### Use case: 
We rely on telegraf heavily to poll MSSQL perf metrics and in case of any errors with polling we would like to capture errors and send it to downstream services for processing.

**Comment by srebhan** [comment]
@batistado for clarification. Do you want to have this `OnError` callback to live in each plugin or do you want this to be global over all plugins?

**Comment by batistado** [comment]
@srebhan I would recommend you have support for both but make it optional. So those who want to get notified on every error for any plugin, they can configure global error handler and those who want it from a specific set of plugins can configure handlers on those plugins.

**Comment by srebhan** [comment]
@batistado from what I understand, you want to do something like this (pseudocode)
```golang
    func myErrorHandler(name string, err error) {
        // Send the received error somewhere
       ...
    }

    ...

    for _, input := range config.RunningInputs {
        if filter.Match(input.LogName()) {
            input.RegisterErrorHandler(myErrorHandler)
        }
    }
    // Same for all other plugin types...
```

Is this correct?

**Comment by srebhan** [comment]
@batistado one more comment. I think this can be done already. If you use the `tail` plugin to monitor your log file, you can easily parse errors (or warnings etc) and then use an output plugin (e.g. http) to push that error/warning/whatsoever somewhere. Would that work for you?

**Comment by batistado** [comment]
> @batistado from what I understand, you want to do something like this (pseudocode)
> 
> ```go
>     func myErrorHandler(name string, err error) {
>         // Send the received error somewhere
>        ...
>     }
> 
>     ...
> 
>     for _, input := range config.RunningInputs {
>         if filter.Match(input.LogName()) {
>             input.RegisterErrorHandler(myErrorHandler)
>         }
>     }
>     // Same for all other plugin types...
> ```
> 
> Is this correct?

Yeah something of this sort.

**Comment by batistado** [comment]
> @batistado one more comment. I think this can be done already. If you use the `tail` plugin to monitor your **Telegraf log file**, you can easily parse errors (or warnings etc) and then use an output plugin (e.g. http) to push that error/warning/whatsoever somewhere. Would that work for you?

Parsing log error messages is a bad developer experience. Also, the code will have to unnecessarily process non error string messages too and figure out if it's error or other kind of log.
And this will also heavily depend on log string message format which could break things if you change your log messages.
I would really only want error messages to be notified.

**Comment by reimda** [comment]
Hi @batistado, I'm not sure I completely understand your use case, but it sounds like you want a way to report errors in an input plugin, like when the sqlserver input fails to connect. You need those errors to be reported downstream through telegraf's output plugin so an external process can take action to fix the errors.

It seems like it should be possible to do this with internal statistics.

https://github.com/influxdata/telegraf/blob/master/plugins/inputs/internal/README.md

see the section about "internal_<plugin_name>"

If the sqlserver doesn't currently report connection errors through internal stats, that would be easy to add. Then if you enable the internal input, telegraf will report the errors through a metric like "internal_sqlserver connection_errors=1" You could watch for that field to increase and take action on it.

**Comment by srebhan** [comment]
@reimda I think the internal plugin is too coarse. From what I understand @batistado wants to know the kind of error (not only the count over all types) to e.g. restart some services on connection errors etc.

**Comment by Hipska** [comment]
Yes, the metrics can be sent to a http or socket (tcp/UDP) port or via an external tool/script with exec(d) plugin.

**Comment by goswamisandeep** [comment]
We do have a scenario where we do report to user whether metric collection fails because of :- 

1. TCP connection issue 
2. Username / password is wrong 
3. PERMISSION issue

I do see telegraf do report them in log files , 

`2022-11-12T09:42:09Z E! [inputs.sqlserver] Error in plugin: query SQLServerAvailabilityReplicaStates failed for server: 10.53.117.51 and database: <empty-database-name> with Msg 18456, Level 14, State 1:, Line 1, Error: mssql: login error: Login failed for user 'sa'.`

is there a way to propogate them back to use

**Comment by Hipska** [comment]
This is also a valid use-case that would be solved with this feature request. Currently it is not possible, but we are looking for help for someone to implement this.
